fairscale/experimental/nn/ssd_offload.py

```diff
@@ -237,61 +245,129 @@ def unwrap(e: Any) -> torch.Tensor:
+        byte_iter = None
+        filename = self.filename
+        if self.override_directory_path is not None:
+            head, tail = os.path.split(self.filename)
+            filename = os.path.join(self.override_directory_path, tail)
+            print(f"overridding path: {os.path.join(head, tail)} to {filename}")
+        if self.is_available():
+            byte_iter = iter(TensorChunkingIterator(self.tensor))
+        else:
+            byte_iter = iter(
+                FileChunkingIterator(self.filename, expected_size_bytes=self.numel() * self.element_size())
+            )
+        return (
+            self.__unpickle__,  # Callable
+            # Args to the callable above
+            (self._shape, self._dtype, self.requires_grad, filename),
+            None,
+            byte_iter,
+        )
+
+    def append(self, item: bytes) -> None:
+        assert self._unpickle_f
+        self._unpickle_f.write(item)
+
+    def extend(self, items: List[bytes]) -> None:
+        for i in items:
+            self.append(i)
+
+
+class CheckpointPathContextManager:
+    def __init__(self, override_path: str) -> None:
+        self.old_path = SsdTensorHandle.override_directory_path
+        self.override_path = override_path
+
+    def __enter__(self) -> None:
+        SsdTensorHandle.override_directory_path = self.override_path
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        exec_traceback: Optional[TracebackType],
+    ) -> None:
+        SsdTensorHandle.override_directory_path = self.old_path
+
+
+# Classes supporting torch.save/load
+class TorchSaver:
+    def __init__(self) -> None:
+        self.pickle_module = DisableMemoizationPicklerModule
+
+    def save(
+        self, obj: Any, f: Union[str, os.PathLike, BinaryIO, IO[bytes]], pickle_protocol: int = DEFAULT_PROTOCOL
+    ) -> None:
+        torch.serialization.save(
+            obj, f, self.pickle_module, pickle_protocol=pickle_protocol, _use_new_zipfile_serialization=False
+        )
+
+
+class SsdParameter(SsdTensorHandle, torch.nn.Parameter):
+    @classmethod
+    def from_tensor(cls: Type[SsdParameter], tensor: SsdTensorHandle) -> SsdParameter:  # type: ignore
+        r = cls(tensor.shape, tensor.dtype, tensor.requires_grad)
+        r.point_to_tensor(tensor)
+        return r
 
-class SsdFlatParameter(torch.nn.Parameter, SsdTensorHandle):
+    @staticmethod
+    def __new__(
+        cls: Type[SsdParameter], shape: torch.Size, dtype: torch.dtype, requires_grad: bool = True
+    ) -> SsdParameter:
+        r = super(SsdParameter, cls).__new__(cls, shape, dtype=dtype, requires_grad=requires_grad)
+        return r  # type: ignore
+
+    def __init__(self, shape: torch.Size, dtype: torch.dtype, requires_grad: bool = True) -> None:
+        super(SsdParameter, self).__init__(shape, dtype, requires_grad)
+
+
+class SsdFlatParameter(SsdParameter):
```

Comment:
just wondering if we can always use SsdFlatParameter since it is a generic version of the FlatParameter. Thats probably a TODO for the future.

Reply:
I'm not sure, right now SsdFlatParameters must have their file_params (filename and offset). In addition the SsdTensorHandle hook will automatically write back the tensor to disk whenever there is a modification. I don't think we want that default behavior when FlatParameters are stored inside of GPU/CPU memory.